### PR TITLE
Disable InvalidateVisual when Visual's not visible

### PR DESCRIPTION
**Pull request by ltetak**
## What does the pull request do?
Alternative to https://github.com/AvaloniaUI/Avalonia/pull/14684 - does not disable the animation but just prevents it from rendering the visual. There are many edge cases and even more InvalidateVisual calls that might be optimized. This is meant just as a draft for future work.

## What is the current behavior?
It invokes the rendering in all cases.

## What is the updated/expected behavior with this PR?
Changing property or render transform property should not fire rendering if the control is not visible.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

### CODE REVIEW

File: src/Avalonia.Base/Visual.cs [hunk 1]
```diff
@@ -409,7 +409,10 @@ protected static void AffectsRender<T>(params AvaloniaProperty[] properties)
                 {
                     if (e.Sender is T sender)
                     {
-                        sender.InvalidateVisual();
+                        if (sender.IsVisible || e.Property == IsVisibleProperty)
+                        {
+                            sender.InvalidateVisual();
+                        }
```
Comment (maxkatz6) [inline src/Avalonia.Base/Visual.cs:415]: In many controls we call InvalidateVisual directly from OnPropertyChanged, avoiding AffectsRender

File: src/Avalonia.Base/Visual.cs [hunk 2]
```diff
@@ -655,7 +658,10 @@ private static void ZIndexChanged(AvaloniaPropertyChangedEventArgs e)
         /// <param name="e">The event args.</param>
         private void RenderTransformChanged(object? sender, EventArgs e)
         {
-            InvalidateVisual();
+            if (ClipToBounds && IsEffectivelyVisible)
```
Comment (maxkatz6) [inline src/Avalonia.Base/Visual.cs:661]: ClipToBounds?
IsEffectivelyVisible essentially inherits IsVisible property from all parents, combined. It's not related to the clipping or fact of control being displayed on the screen right or not.

**Comment by maxkatz6** [comment]
In general, disabling invalidation for invisible controls makes sense for me, and can be useful as a default behavior. But it might need to be done on compositor layer. cc @kekekeks for his opinion.

**Comment by kekekeks** [comment]
Visual not being effectively visible doesnt prevent it from being used as something like a visual brush (which we want to move to compositor side) or as a composition animation expression target.

**Pull request closed without merging**